File: bilateral.cpp
#include <iostream>
#include <random>
#include <vector>
#include <chrono>
#include <cmath>

int main(int argc, char **argv) {
    int N = 10'000;
    float sigma = 1;
    int radius = 3;
    if (argc > 1) {
        N = std::stoi(argv[1]);
    }
    if (argc > 2) {
        radius = std::stoi(argv[2]);
    }
    if (argc > 3) {
        sigma = std::stof(argv[3]);
    }

    std::vector<float> array(N * N), results(N * N);
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_real_distribution<> dis(0, 1.0);
    int sample_size = N * N;
    for (auto& val : array) {
        val = dis(gen);
    }
    auto start = std::chrono::high_resolution_clock::now();
    size_t total_count = 0;

    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            const int imax = std::min(i + radius, N - 1);
            const int imin = std::max(i - radius, 0);
            const int jmax = std::min(j + radius, N - 1);
            const int jmin = std::max(j - radius, 0);
            const int index_global = N * i + j;
            const float center_val = array[i * N + j];
            float sum_weights = 0.0f;
            float sum_weighted = 0.0f;

            for (int ii = imin; ii <= imax; ++ii) {
                for (int jj = jmin; jj <= jmax; ++jj) {
                    total_count++;
                    int index_local = N * ii + jj;
                    const float diff = center_val - array[index_local];
                    const float w = std::exp(-diff * diff / sigma / sigma / 2);
                    sum_weights += w;
                    sum_weighted += array[index_local] * w;
                }
            }
            results[index_global] = sum_weighted/ sum_weights;
        }
    }
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::cout << "CPU Time: " << duration.count() << " ms" << std::endl;
    std::cout << "Average " << std::accumulate(results.begin(), results.end(), 0.0) / sample_size << std::endl;
    std::cout << "Total Count " << total_count << std::endl;
    return 0;
}

File: bilateral.f90
subroutine str2int(str, int, stat)
    implicit none
    character(len = *), intent(in) :: str
    integer, intent(out) :: int
    integer, intent(out) :: stat

    read(str, *, iostat = stat)  int
end subroutine str2int

subroutine str2real(str, float, stat)
    implicit none
    character(len = *), intent(in) :: str
    real, intent(out) :: float
    integer, intent(out) :: stat

    read(str, *, iostat = stat)  float
end subroutine str2real

program main
    implicit none
    character(len = 256) :: arg_value ! Character variable to store the argument
    integer :: N, count, i, j, status, radius, imax, imin, jmin, jmax, iimax, iimin, jjmin, jjmax
    integer(kind=8) :: total_count
    real, allocatable :: array(:, :), result(:, :), weight(:, :)
    real :: sigma
    real :: start_time, end_time
    count = command_argument_count()
    call RANDOM_SEED()
    radius = 1
    N = 1000
    sigma = 1.0
    do i = 1, count
        call get_command_argument(i, arg_value, status = status)
        if (status == 0) then
            if(i == 1) then
                call str2int(trim(arg_value), N, status)
            end if
            if(i == 2) then
                call str2int(trim(arg_value), radius, status)
            end if
            if(i == 3) then
                call str2real(trim(arg_value), sigma, status)
            end if
            if (status /= 0) then
                print *, "Error parsing argument", i
                STOP 1
            end if
        else
            print *, "Error getting argument", i
            STOP 1
        end if
    end do

    allocate(array(N, N))
    allocate(result(N, N))
    allocate(weight(2 * radius + 1, 2 * radius + 1))
    call RANDOM_NUMBER(array)
    CALL CPU_TIME(start_time)
    total_count = 0
    do i = 1, N
        do j = 1, N
            imin = max(i - radius, 1)
            imax = min(i + radius , N)
            jmin = max(j - radius, 1)
            jmax = min(j + radius , N)
            iimax = radius + imax - i + 1
            iimin = radius + imin - i + 1
            jjmax = radius + jmax - j + 1
            jjmin = radius + jmin - j + 1
            weight(iimin:iimax, jjmin:jjmax) = exp(-(array(imin:imax, jmin:jmax) - array(i, j))**2 / 2 / sigma**2)
            result(i, j) = sum(array(imin:imax, jmin:jmax) * weight(iimin:iimax, jjmin:jjmax))
            result(i, j) = result(i, j) / sum(weight(iimin:iimax, jjmin:jjmax))
            total_count = total_count + (jmax - jmin + 1) * (imax - imin + 1)
        end do
    end do

    CALL CPU_TIME(end_time)
    PRINT *, "CPU Time:", (end_time - start_time) * 1000, "ms"
    print *, "Average", sum(result)/N/N
    print *, "Total Count", total_count
end